src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -55,6 +62,38 @@ internal bool AddGas(long gas)
             return testMode || GasConsumed <= gas_amount;
         }
 
+        internal void CallFromNativeContract(Action onComplete, UInt160 hash, string method, params StackItem[] args)
+        {
+            invocationStates.Add(CurrentContext, new InvocationState
+            {
+                ReturnType = typeof(void),
+                Callback = onComplete
+            });
+            CallContract(hash, method, new VMArray(args));
+        }
+
+        internal void CallFromNativeContract<T>(Action<T> onComplete, UInt160 hash, string method, params StackItem[] args)
+        {
+            invocationStates.Add(CurrentContext, new InvocationState
+            {
+                ReturnType = typeof(T),
+                Callback = onComplete
+            });
+            CallContract(hash, method, new VMArray(args));
+        }
+
+        protected override void ContextUnloaded(ExecutionContext context)
+        {
+            base.ContextUnloaded(context);
+            if (!(UncaughtException is null)) return;
+            if (invocationStates.Count == 0) return;
+            if (!invocationStates.TryGetValue(CurrentContext, out InvocationState state)) return;
```

Comment:
It can't be `null`. I have checked the `invocationStates.Count` in L89.

Reply:
It could be `null`, when execute `ret` of the last context, it'll set `CurrentContext` to `null` before `ContextUnloaded`. And L89 will not throw exception.

Reply:
if so, `invocationStates.Count` must be `0` and it will return in L89.

Reply:
But we use `invocationStates.TryGetValue` in L90, should change to `invocationStates.TryPopValue`?

Reply:
```suggestion
            if (!invocationStates.Remove(CurrentContext, out InvocationState state)) return;
```

Reply:
It should be `Remove()`. I fixed it.